File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;
class Solution {
    public int solution(String dirs) {
        ArrayList<String> list = new ArrayList<>();
        int answer = 0;
        int[] pos = {0, 0};
        char[] c = dirs.toCharArray();
        for(int i = 0; i < c.length; i++) {
            String currentPos = pos[0] + "" + pos[1];
            if(c[i] == 'U' && pos[1] < 5) {
                pos[1]++;
            }
            else if(c[i] == 'D' && pos[1] > -5) {
                pos[1]--;
            }
            else if (c[i] == 'L' && pos[0] > -5) {
                pos[0]--;
            }
            else if (c[i] == 'R' && pos[0] < 5){
                pos[0]++;
            }
            String movPos = pos[0] + "" + pos[1];
            String chk1 = currentPos + movPos;
            String chk2 = movPos + currentPos;
          
            if(!list.contains(chk1) && !list.contains(chk2) && !currentPos.equals(movPos)) {
                answer++;
                list.add(chk1);
                System.out.println(chk1);
                list.add(chk2);
            }
        }
      
        return answer;
    }
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    var answer = 0;
    const command = {
        'U' : [0,1],
        'D' : [0,-1],
        'R' : [1,0],
        'L' : [-1,0]
    }
    const set = new Set();
    function isValid(y,x) {
        return x >= -5 && x <= 5 && y >= -5 && y <= 5;
    }
    const pos = [0,0];
    for (const dir of dirs) {
        const nextY = pos[1] + command[dir][1];
        const nextX = pos[0] + command[dir][0];
        if (!isValid(nextY,nextX)) continue;
        const path1 = `${pos[0]}${pos[1]}${nextX}${nextY}`;
        const path2 = `${nextX}${nextY}${pos[0]}${pos[1]}`;
        if (!set.has(path1) && !set.has(path2)) {
            set.add(path1);
            set.add(path2);
            answer++;
        }
        pos[0] = nextX;
        pos[1] = nextY;
    
    }
    
    return answer;
}